weave/trace_server/opentelemetry/attributes.py

```diff
@@ -313,199 +308,90 @@ def unflatten_key_values(
+}
 
-    def get_weave_outputs(self) -> Any:
-        outputs: dict[str, Any] = (
-            self.get_attribute_value(oi.SpanAttributes.LLM_OUTPUT_MESSAGES) or {}
-        )
-        result = {}
-        for k, v in outputs.items():
-            if k.isdigit() and isinstance(v, dict):
-                for key in v.keys():
-                    result[key + f"_{k}"] = v[key]
-        return to_json_serializable(result)
-
-    def get_weave_inputs(self) -> Any:
-        inputs: dict[str, Any] = (
-            self.get_attribute_value(oi.SpanAttributes.LLM_INPUT_MESSAGES) or {}
-        )
-        result = {}
-        for k, v in inputs.items():
-            if k.isdigit() and isinstance(v, dict):
-                for key in v.keys():
-                    result[key + f"_{k}"] = v[key]
-        return to_json_serializable(result)
-
-    def get_weave_attributes(
-        self, extra: Optional[dict[str, Any]] = None
-    ) -> dict[str, Any]:
-        system = self.get_attribute_value(oi.SpanAttributes.LLM_SYSTEM)
-        provider = self.get_attribute_value(oi.SpanAttributes.LLM_PROVIDER)
-        invocation_parameters = self.get_attribute_value(
-            oi.SpanAttributes.LLM_INVOCATION_PARAMETERS
-        )
-        kind = self.get_attribute_value(oi.SpanAttributes.OPENINFERENCE_SPAN_KIND)
-        model = self.get_attribute_value(oi.SpanAttributes.LLM_MODEL_NAME)
-        attributes = {
-            "system": str(system) if system else None,
-            "provider": str(provider) if provider else None,
-            "kind": str(kind) if kind else None,
-            "model": str(model) if model else None,
-        }
+for key in (
+    USAGE_KEYS["prompt_tokens"]
+    + USAGE_KEYS["completion_tokens"]
+    + USAGE_KEYS["total_tokens"]
+):
+    KEY_HANDLERS[key] = try_parse_int
 
-        if invocation_parameters:
-            try:
-                js = json.loads(invocation_parameters)
-                for k, v in js.items():
-                    attributes[k] = str(v)
-            except json.JSONDecodeError:
-                raise ValueError(f"Invalid JSON string: {invocation_parameters}")
 
-        if extra:
-            attributes.update(extra)
+class SpanEvent(dict):
+    name: str
+    timestamp: datetime
+    attributes: dict[str, Any]
+    dropped_attributes_count: int
 
-        return to_json_serializable(attributes)
 
+def parse_weave_values(
+    attributes: dict[str, Any], key_mapping: dict[str, list[str]]
+) -> dict[str, Any]:
+    result = {}
+    for key, attribute_key_list in key_mapping.items():
+        for attribute_key in attribute_key_list:
+            value = get_attribute(attributes, attribute_key)
+            if value:
+                if attribute_key in KEY_HANDLERS:
+                    try:
+                        value = KEY_HANDLERS[attribute_key](value)
+                    except:
+                        pass
+                result[key] = value
+                break
+    return result
```

Comment:
`to_json_serializable` here just to be safe. Haven't been able to trigger a json serialization error yet but would be bad if somehow got triggered.